Accordion

ahuth opened the pull request:
https://github.com/juanca/react-aria-components/projects/1#card-10586797
https://www.w3.org/TR/wai-aria-practices-1.1/#accordion

Accordion component.

Everything closed:
<img width="1035" alt="closed" src="https://user-images.githubusercontent.com/2503289/43695082-5d8fd4ac-98eb-11e8-913d-558c491b37b7.png">

One section open:
<img width="1035" alt="open" src="https://user-images.githubusercontent.com/2503289/43695090-62ff7046-98eb-11e8-932a-0cd60670535a.png">

Commit:
Turn off some annoying lint rules

Signed-off-by: Jason Steck <jsteck@mavenlink.com>

jsteckml committed:
Add basic <Accordion> component

Signed-off-by: Andrew Huth <ahuth@mavenlink.com>

Commit:
Implement every accordion header as a button

Signed-off-by: Jason Steck <jsteck@mavenlink.com>

Commit:
Remove unused file

Commit:
Alphabetize elintrc

Commit:
Wrap the accordion header and panel in a <div>

Commit:
Link up the accordion header and panel content via an ID

Commit:
Allow the accordion header button to be marked as disabled

This happens when the panel is open, and we do not want to allow all panels to be closed. Meaning, there has to be at least one panel open. In that case, the open header button must be marked as aria-disabled.

Commit:
Alphabetize prop types

ahuth commented on .eslintrc:
```diff
@@ -2,8 +2,10 @@
     "prefer-destructuring": "off",
     "react/destructuring-assignment": "off",
     "react/jsx-filename-extension": "off",
+    "react/jsx-one-expression-per-line": "off"
```
We probably don't want to keep adding to this file, but these rules are pretty annoying.

Commit:
Highlight custom accordion titles in the accordion example

Commit:
Add basic accordion styling

Commit:
Use custom title for all accordion examples

Commit:
Add some styling on focus/hover of the accordion

Commit:
Use heading tags in the accordion to get proper semantics

ahuth commented on src/accordion/section.js:
```diff
@@ -0,0 +1,85 @@
+
+    if (!Heading) {
+      throw new Error('Accordion must have a headingLevel between 1 and 6');
+    }
```
I'm not a huge fan of runtime checks. But in this case, it's super important from an accessibility standpoint that we get the correct "heading" level. Figured it was worth it here to break fast and loud if someone doesn't set that up with something valid.

jsteckml replied:
^ agreed

ahuth commented on src/accordion/section.js:
```diff
@@ -0,0 +1,85 @@
+          >
+            {titleVal}
+          </button>
+        </Heading>
```
Note the structure here of:

```jsx
<Heading>
  <Button />
</Heading>
```

Part of me wants to make this whole sequence a render prop and let users set whatever DOM they want. 

However, [the authoring practices](https://www.w3.org/TR/wai-aria-practices-1.1/#accordion) are clear that the only child of the "heading" should be the "button", and it made more sense to me to be strict.

ahuth commented:
@juanca, can you take a look at this and let me know what you think?

cc @jsteckml

ahuth commented on src/accordion/accordion.css:
```diff
@@ -0,0 +1,5 @@
+.container {
+  border: 1px solid LightGrey;
+  border-radius: 3px;
+  box-shadow: 0 1px 2px LightGrey;
```
Heh, probably don't want to have drop shadow on here that someone will have to take off. Was following the WAI-ARIA examples styling.

ahuth commented on src/accordion/section.js:
```diff
@@ -0,0 +1,85 @@
+}
+
+Section.propTypes = {
+  cannotClose: PropTypes.bool,
```
`cannotClose` is an interesting one, and maybe there's a better way to word or implement it.

Some accordions must have at least panel open (ie - you can't close all of them). In that case, the one that's open must have `aria-disabed="true"` set on its "button".

jsteckml replied:
I think `cannotClose` is fine. Alternatively we could call it `disabled` and/or also have it add the disabled attribute to the element

jsteckml commented on src/accordion/section.js:
```diff
@@ -0,0 +1,85 @@
+    const Heading = headings[headingLevel];
+
+    if (!Heading) {
+      throw new Error('Accordion must have a headingLevel between 1 and 6');
```
Fun fact: if a user passes in `'toString'` it'll get past this check (but will still fail proptypes) and it'll even render something (`[object Object]` probably?)

And if they reaaaally wanted to, they could:
```jsx

Object.prototype.Cheat = function Cheat() {
  return <div> I hate accessible headings and buttons </div>
};

<Section headingLevel='Cheat' ...>
```

Just thought that was an funny work around 😜 I don't think we need to account for that at all.

ahuth replied:
Really interesting point, though. I forgot that objects have some other properties on them because of the prototype.

Seems like a `hasOwnProperty` check instead wouldn't be that bad, and would prevent this edge case.

jsteckml commented on src/examples/accordion-example.js:
```diff
@@ -0,0 +1,73 @@
+        <Accordion>
+          <Section
+            headingLevel={3}
+            onClick={this.handleClick('first')}
```
Just realized this function returns a new function on every render. We could have created a `handleClickFirst = () => { this.handleClick('first') }` but performance probably doesn't matter in this example.

ahuth replied:
Yeah, you're right. A better or real-world example would probably want to have separate handlers for each section, as you've suggested.

👍

Commit:
Use "hasOwnProperty" to deal with edge cases

Do not want someone to use pass in as the "headingLevel" prop an Object prototype property, such as "toString".

Commit:
Merge branch 'master' into accordion

Commit:
Remove styles that users would have to override

Commit:
Whitespace

ahuth commented:
@juanca unless you have any reservations, I'm planning on merging this thing later today. Happy to revisit anything on here, as well.

ahuth commented:
Mergin'

juanca commented:
LGTM for posterity's sake.
